[PyTorch] Take advantage of string literals in TORCH_WARN

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54032 [PyTorch] Take advantage of string literals in TORCH_WARN**

Add a `const char*` override to c10::Warning::warn so that we can avoid wrapping plain C string literals in std::string.

Differential Revision: [D27061983](https://our.internmc.facebook.com/intern/diff/D27061983/)

swolchok committed:
[PyTorch] Take advantage of string literals in TORCH_WARN

Add a `const char*` override to c10::Warning::warn so that we can avoid wrapping plain C string literals in std::string.

Differential Revision: [D27061983](https://our.internmc.facebook.com/intern/diff/D27061983/)

[ghstack-poisoned]

mruberry commented on c10/util/Exception.cpp:
```diff
@@ -133,6 +133,10 @@ void warn(SourceLocation source_location, const std::string& msg, const bool ver
 }
 
+void warn(SourceLocation source_location, const char* msg, const bool verbatim) {
+  ThreadWarningHandler::get_handler()->process(source_location, msg, verbatim);
```
But doesn't this convert to a std::string, anyway?

ezyang replied:
The point is that the std::string constructor only gets code generated once (in the Exception.cpp), rather than at every string literal call site.

swolchok committed:
Update on "[PyTorch] Take advantage of string literals in TORCH_WARN"


Add a `const char*` override to c10::Warning::warn so that we can avoid wrapping plain C string literals in std::string.

Differential Revision: [D27061983](https://our.internmc.facebook.com/intern/diff/D27061983/)

[ghstack-poisoned]

swolchok committed:
unbreak test on "[PyTorch] Take advantage of string literals in TORCH_WARN"


Add a `const char*` override to c10::Warning::warn so that we can avoid wrapping plain C string literals in std::string.

Differential Revision: [D27061983](https://our.internmc.facebook.com/intern/diff/D27061983/)

[ghstack-poisoned]

mruberry reviewed (approved):
Stamped!
